[HUDI-1623] Solid completion time on timeline

boneanxs opened the pull request:
### Change Logs

_Describe context and summary for this change. Highlight if any code was copied._
1. Allow `HoodieInstant` to identify complete instants with completion time
2. Name change: `stateTransitionTime` -> `completionTime`
3. Add new interface `TimeGenerator` and one impl `WaitBasedTimeGenerator`. The singleton `TimeGenerators` is used to manage it.
4. For write client and meta client, we all introduce public APIs for new instant time generation
5. Passing some configs to the `metaClient`, so the active timeline from the meta client could do metadata file state transitions with the time generator.
6. Introduce `HoodieTimeGeneratorConfig` and by default, we have a in-process lock provider based `TimeGenerator` impl.

The complete instant will add completion time, for example,

```java
drwxr-xr-x  2 hui.an  staff    64B Sep  5 18:34 archived
-rw-r--r--  1 hui.an  staff   784B Sep  5 18:34 hoodie.properties
-rw-r--r--  1 hui.an  staff     0B Sep  5 18:35 20230905183414319.commit.requested
-rw-r--r--  1 hui.an  staff   2.1K Sep  5 18:36 20230905183414319.inflight
-rw-r--r--  1 hui.an  staff   5.3K Sep  5 18:37 20230905183414319_20230905183706849.commit
```
### Impact

_Describe any public API or user-facing feature change or any performance impact._
none

### Risk level (write none, low medium or high below)

_If medium or high, explain what verification was done to mitigate the risks._
medium

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change_

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/PreferWriterConflictResolutionStrategy.java:
```diff
@@ -104,7 +104,7 @@ private Stream<HoodieInstant> getCandidateInstantsForTableServicesCommits(Hoodie
     // Merge and sort the instants and return.
     List<HoodieInstant> instantsToConsider = Stream.concat(completedCommitsStream, inflightIngestionCommitsStream)
-        .sorted(Comparator.comparing(o -> o.getStateTransitionTime()))
+        .sorted(Comparator.comparing(o -> o.getCompletionTime()))
```
Only completed instant has completion time?

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieClient.java:
```diff
@@ -95,6 +98,8 @@ protected BaseHoodieClient(HoodieEngineContext context, HoodieWriteConfig client
     this.metrics = new HoodieMetrics(config);
     this.txnManager = new TransactionManager(config, fs);
+    TrueTimeGeneratorManager.initialize(
+        new HoodieTrueTimeConfig(new TypedProperties(clientConfig.getProps())), hadoopConf);
```
Kind of think we should have a tool method in `HoodieWriteConfig` named `HoodieWriteConfig#getTimeGeneratorConfig`, and let's rename `TrueTimeGeneratorManager` -> `TimeGeneratorManager`

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -200,7 +200,8 @@ public class HoodieWriteConfig extends HoodieConfig {
       .defaultValue(Integer.toString(TimelineLayoutVersion.CURR_VERSION))
-      .withValidValues(Integer.toString(TimelineLayoutVersion.VERSION_0), Integer.toString(TimelineLayoutVersion.VERSION_1))
+      .withValidValues(Integer.toString(TimelineLayoutVersion.VERSION_0),
+          Integer.toString(TimelineLayoutVersion.VERSION_1), Integer.toString(TimelineLayoutVersion.VERSION_2))
```
Do we need a VERSION_2 timeline layout? The layout does not change actually, we only change the fine name convention of completed metadata files.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -197,19 +204,28 @@ public HoodieActiveTimeline() {
+  private boolean shouldIncludeCompletionTime(HoodieInstant instant) {
+    return instant.isCompleted()
+        && metaClient.getTimelineLayoutVersion().getVersion() == TimelineLayoutVersion.VERSION_2;
+  }
```
Do we need this decision

```java
metaClient.getTimelineLayoutVersion().getVersion() == TimelineLayoutVersion.VERSION_2
```

The old timeline does not even have `VERSION_2` in the code path.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -600,27 +617,37 @@ private void transitionState(HoodieInstant fromInstant, HoodieInstant toInstant,
-        Path fromInstantPath = getInstantFileNamePath(fromInstant.getFileName());
-        Path toInstantPath = getInstantFileNamePath(toInstant.getFileName());
+        createFileInMetaPath(fromInstantFileName, data, allowRedundantTransitions);
+        toInstantFileName = instantFormat.getInstantFileName(toInstant);
```
Do we need to change this very old timeline layput specifically here? Should we handle it in the upgrade handler?

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -795,6 +822,22 @@ protected void createFileInMetaPath(String filename, Option<byte[]> content, boo
+    try {
+      trueTimeGenerator.lock();
+      String currentTime = HoodieInstantTimeGenerator.formatDate(new Date(trueTimeGenerator.nowAsMillis(false)));
+      HoodieInstant instantWithCompletionTime = new HoodieInstant(instant.getState(), instant.getAction(),
```
Can you explain why the `HoodieActiveTimeline#createNewInstantTime` can not work correctly here?

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -795,6 +822,22 @@ protected void createFileInMetaPath(String filename, Option<byte[]> content, boo
     }
   }
+  
+  protected void createCompleteInstantWithCompletionTime(HoodieInstant instant, Option<byte[]> content) {
```
rename to `createCompleteFileInMetaPath`

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -955,8 +956,9 @@ public static HoodieTableFileSystemView getFileSystemView(HoodieTableMetaClient
     HoodieTimeline timeline = metaClient.getActiveTimeline();
     if (timeline.empty()) {
+      HoodieTimeGeneratorConfig timeGeneratorConfig = HoodieTimeGeneratorConfig.newBuilder().withPath(metaClient.getBasePathV2().toString()).build();
       final HoodieInstant instant = new HoodieInstant(false, HoodieTimeline.DELTA_COMMIT_ACTION,
```
The write config/lock provider config does not pass around correctly, not sure whether we should take a lock here as this is a metadata table.

boneanxs replied:
This method is just an initialize step, right? If so, I think the instantTime generated here is just a fake time, and won't be used actually.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1406,7 +1408,10 @@ public static String deleteMetadataTable(HoodieTableMetaClient dataMetaClient, H
+      HoodieTimeGeneratorConfig timeGeneratorConfig = HoodieTimeGeneratorConfig.newBuilder()
+          .withPath(dataMetaClient.getBasePathV2().toString()).build();
+      final Path metadataBackupPath = new Path(metadataTablePath.getParent(),
+          ".metadata_" + HoodieActiveTimeline.createNewInstantTime(timeGeneratorConfig, dataMetaClient.getHadoopConf()));
```
We can ignore the lock.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1462,8 +1467,11 @@ public static String deleteMetadataTablePartition(HoodieTableMetaClient dataMeta
       final Path metadataPartitionBackupPath = new Path(metadataTablePartitionPath.getParent().getParent(),
-          String.format(".metadata_%s_%s", partitionType.getPartitionPath(), HoodieActiveTimeline.createNewInstantTime()));
+          String.format(".metadata_%s_%s", partitionType.getPartitionPath(),
+              HoodieActiveTimeline.createNewInstantTime(timeGeneratorConfig, dataMetaClient.getHadoopConf())));
```
We can ignore the lock.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimeGeneratorBase.java:
```diff
@@ -0,0 +1,130 @@
+  private final HoodieTimeGeneratorConfig config;
+  private final LockConfiguration lockConfiguration;
+
+  private final Configuration hadoopConf;
```
Should use `SerializableConfiguration hadoopConf` instead.

danny0405 commented on hudi-common/src/test/java/org/apache/hudi/common/table/timeline/TestHoodieInstantTimeGenerator.java:
```diff
@@ -0,0 +1,41 @@
+  private static HoodieTimeGeneratorConfig timeGeneratorConfig = HoodieTimeGeneratorConfig.newBuilder()
+      .withPath("/tmp").build();
+  private static Configuration hadoopConf = new Configuration();
+
```
Remove it or make it static final.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieTimeGeneratorConfig.java:
```diff
@@ -0,0 +1,113 @@
+
+import static org.apache.hudi.common.config.LockConfiguration.LOCK_PREFIX;
+
+public class HoodieTimeGeneratorConfig extends HoodieConfig {
```
We should refactor this class to be together with other config classes, can you fire a JIRA issue to trace this?

boneanxs replied:
Move to `org.apache.hudi.common.config` for now

danny0405 commented on hudi-cli/src/main/java/org/apache/hudi/cli/commands/CompactionCommand.java:
```diff
@@ -195,9 +196,10 @@ public String scheduleCompact(
     // First get a compaction instant time and pass it to spark launcher for scheduling compaction
-    String compactionInstantTime = HoodieActiveTimeline.createNewInstantTime();
+    String compactionInstantTime = HoodieActiveTimeline.createNewInstantTime(timeGeneratorConfig, HoodieCLI.conf);
 
```
Kind of think the invoke should be:

`TimeGenerators.getInstance(HoodieCLI.basePath, HoodieCLI.conf).createNewInstantTime`

boneanxs replied:
Fire the ticket to track this refactor: [HUDI-6860](https://issues.apache.org/jira/browse/HUDI-6860)

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/CompactHelpers.java:
```diff
@@ -82,7 +82,7 @@ public HoodieCommitMetadata createCompactionMetadata(
     HoodieActiveTimeline activeTimeline = table.getActiveTimeline();
     try {
-      activeTimeline.transitionCompactionInflightToComplete(
+      activeTimeline.transitionCompactionInflightToComplete(false, table.getConfig().getTimeGeneratorConfig(),
```
Are we sure there is no need to lock here?

boneanxs replied:
Yes, the outside caller use transitionManager to guarantee this.

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/rollback/BaseRollbackActionExecutor.java:
```diff
@@ -261,7 +261,7 @@ protected void finishRollback(HoodieInstant inflightInstant, HoodieRollbackMetad
       // to complete in the data table timeline
       if (!skipTimelinePublish) {
-        table.getActiveTimeline().transitionRollbackInflightToComplete(inflightInstant,
+        table.getActiveTimeline().transitionRollbackInflightToComplete(false, table.getConfig().getTimeGeneratorConfig(), inflightInstant,
```
false or `!enableLocking` ?

boneanxs replied:
Added comments here

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -304,9 +326,23 @@ protected void deleteInstantFile(HoodieInstant instant) {
+   * timeline to get completionTime
+   * TODO Depreciate this method and fix relate tests
+   */
+  private String getInstantFileName(HoodieInstant instant) {
```
You mean the `instant` passed in is constructed manually instead of filtered from the timeline?

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimeGenerator.java:
```diff
@@ -0,0 +1,39 @@
+   * @return Current true time as millis.
+   */
+  long currentTimeMillis(boolean shouldLock);
+
```
We already have `lock` and `unlock` so do we still need the `shouldLock` flag?

boneanxs replied:
refactor the interface to wrap the lock inside.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimeGeneratorType.java:
```diff
@@ -0,0 +1,34 @@
+
+  @EnumFieldDescription("A wait based time generator, holding a mutex lock with skew times to "
+      + "produce globally monotonically increasing timestamp")
+  WAIT_BASED
```
`LOCK_PROVIDER` and `TRUE_TIME`

boneanxs commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieTimeGeneratorConfig.java:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  public static class Builder {
+    private final HoodieTimeGeneratorConfig timeGeneratorConfig = new HoodieTimeGeneratorConfig();
```
Since we need to set default in build() method, we need to create `HoodieTimeGeneratorConfig` first, also, this is compatible with other HoodieConfig

danny0405 commented:
[1623.patch.zip](https://github.com/apache/hudi/files/12607928/1623.patch.zip)
Thanks for the contribution, I have reviewed and applied some changes.

danny0405 commented on hudi-cli/src/test/java/org/apache/hudi/cli/integ/ITTestClusteringCommand.java:
```diff
@@ -95,7 +95,7 @@ public void testScheduleClustering() throws IOException {
 
     // there is 1 requested clustering
-    HoodieActiveTimeline timeline = HoodieCLI.getTableMetaClient().getActiveTimeline();
+    HoodieActiveTimeline timeline = HoodieCLI.getTableMetaClient().getActiveTimeline().reload();
```
Why this change?

danny0405 commented on hudi-cli/src/test/java/org/apache/hudi/cli/integ/ITTestCompactionCommand.java:
```diff
@@ -283,7 +283,7 @@ private String prepareScheduleCompaction() {
 
     // get compaction instance
-    HoodieActiveTimeline timeline = HoodieCLI.getTableMetaClient().getActiveTimeline();
+    HoodieActiveTimeline timeline = HoodieCLI.getTableMetaClient().getActiveTimeline().reload();
```
Why this change?

boneanxs replied:
Removed

danny0405 commented on hudi-cli/src/main/java/org/apache/hudi/cli/commands/ClusteringCommand.java:
```diff
@@ -59,7 +58,7 @@ public String scheduleClustering(
 
     // First get a clustering instant time and pass it to spark launcher for scheduling clustering
-    String clusteringInstantTime = HoodieActiveTimeline.createNewInstantTime();
+    String clusteringInstantTime = client.getActiveTimeline().createNewInstantTime();
```
Kind of think we should avoid to instantiate the active timeline where possible, the active timeline construction actually triggers the metadata files listing, which is not relevent with time generation.

boneanxs replied:
Moved to metaClient

danny0405 commented on hudi-client/hudi-client-common/src/test/java/org/apache/hudi/utils/TestMetadataConversionUtils.java:
```diff
@@ -82,6 +82,7 @@ public void init() throws IOException {
     String newCommitTime = HoodieTestTable.makeNewCommitTime();
     createCleanMetadata(newCommitTime);
+    metaClient.reloadActiveTimeline();
     HoodieArchivedMetaEntry metaEntry = MetadataConversionUtils.createMetaWrapper(
```
Do we still need the reload?

boneanxs replied:
Yes, this is necessary, MetadataConversionUtils.createMetaWrapper will try to get the complete instant path from activeTimeline, but if we don't reloadActiveTimeline, the complete instant written above "createCleanMetadata(newCommitTime);" will not be included in activeTimeline.

boneanxs replied:
Remove other reloadActiveTimeline except `testCompletedCompaction`, since it will initialize activeTimeline when create commits. We have to reload here to ensure new commits are added.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/util/StreamerUtil.java:
```diff
@@ -163,6 +166,27 @@ public static HoodieIndexConfig getIndexConfig(Configuration conf) {
+        .fromProperties(flinkConf2TypedProperties(conf));
+
+    if (OptionsResolver.isLockRequired(conf) && !conf.containsKey(HoodieLockConfig.LOCK_PROVIDER_CLASS_NAME.key())) {
+      // configure the fs lock provider by default
```
Do we need this? Shouldn't the lock provider already be configured in the configuration? Or maybe
`HoodieWriteClients.getHoodieClientConfig(Configuration).getHoodieTimeGeneratorConfig()` should work?

danny0405 commented:
cc @vinothchandar It is great if you could help with the final review. The main change list:

1. for write client and meta client, we all introduce public APIs for new instant time generation
2. we pass some configs to the meta client, so the active timeline from the meta client could do metadata file state transitions with the time generator.
3. Introduces a `TimeGegerator` API
4. Introduces `HoodieTimeGeneratorConfig` and by default, we have a in-process lock provider based `TimeGenerator` impl.

danny0405 commented on hudi-cli/src/main/java/org/apache/hudi/cli/HoodieCLI.java:
```diff
@@ -85,7 +86,9 @@ public static void initFS(boolean force) throws IOException {
+    setTableMetaClient(HoodieTableMetaClient.builder().setConf(HoodieCLI.conf).setBasePath(basePath).setLoadActiveTimelineOnLoad(false)
+        .setConsistencyGuardConfig(HoodieCLI.consistencyGuardConfig)
+        .setTimeGeneratorConfig(HoodieTimeGeneratorConfig.defaultConfig(basePath))
         .setLayoutVersion(Option.of(layoutVersion)).build());
```
Can we use the default config here, I saw some `createNewInstantTime` callings from some of the CLI commands.

boneanxs replied:
Added [HUDI-6884](https://issues.apache.org/jira/browse/HUDI-6884) to track this follow-up.

boneanxs commented on hudi-integ-test/src/test/java/org/apache/hudi/integ/ITTestHoodieDemo.java:
```diff
@@ -115,19 +115,20 @@ public void testParquetDemo() throws Exception {
     testHiveAfterFirstBatch();
     testPrestoAfterFirstBatch();
-    testTrinoAfterFirstBatch();
+    // testTrinoAfterFirstBatch();
```
Comment here for now to see if other tests can pass

boneanxs commented on hudi-platform-service/hudi-metaserver/hudi-metaserver-client/src/main/java/org/apache/hudi/common/table/timeline/HoodieMetaserverBasedTimeline.java:
```diff
@@ -82,4 +88,18 @@ protected Option<byte[]> readDataFromPath(Path detailPath) {
+    if (instant.isCompleted()) {
+      // Set a fake completion time.
+      return instant.getFileName("0").replace("_0", "");
+    }
```
Added [HUDI-6883](https://issues.apache.org/jira/browse/HUDI-6883) to track this follow-up. To add completion time for `HoodieMetaserverBasedTimeline`.

boneanxs commented:
@danny0405 The trino test failure is caused by trino image always use an old hudi-trino-bundle jar:
we can see `/usr/local/trino/plugin/hive/hudi-trino-bundle.jar`

```bash
root@trino-coordinator-1:/usr/local/trino-server-368# ls -larth ${TRINO_HOME}/plugin/hive/hudi-trino-bundle.jar
-rw-r--r-- 1 root root 14M Jan 12  2022 /usr/local/trino/plugin/hive/hudi-trino-bundle.jar
```

but from presto image, the new built jar is used.

```bash
root@presto-coordinator-1:/opt/presto-server-0.271# ls -larth ${PRESTO_HOME}/plugin/hive-hadoop2/hudi-presto-bundle.jar
-rw-r--r-- 1 root root 42M Sep 23 06:04 /opt/presto-server-0.271/plugin/hive-hadoop2/hudi-presto-bundle.jar
```

We need to add explicitly copy command to copy the built jar to `/usr/local/trino/plugin/hive/` just like presto image does in the entrypoint

```bash
# presto entrypoint.sh(docker/hoodie/hadoop/prestobase/bin/entrypoint.sh)

# Copy the presto bundle at run time so that locally built bundle overrides the one that is present in the image
cp ${HUDI_PRESTO_BUNDLE} ${PRESTO_HOME}/plugin/hive-hadoop2/
```

we need to add this also in the trino entrypoint

```bash
#docker/hoodie/hadoop/trinobase/scripts/trino.sh
cp ${HUDI_TRINO_BUNDLE} ${TRINO_HOME}/plugin/hive/
```

Given this pr changed the timeline format, so old jar will fail to parse instants and return wrong results. We may need to change `trino.sh` and rebuild the image whereas the `intergration-tests` can use it.

codope commented:
Sure, i'll fire a fix tonight.

codope commented:
@boneanxs Can you review this PR https://github.com/apache/hudi/pull/9781? When I build the image locally with the changes, it does not udpate the bundle jar in the container. Still trying to figure out why.

boneanxs commented:
> @boneanxs Can you review this PR #9781? When I build the image locally with the changes, it does not udpate the bundle jar in the container. Still trying to figure out why.

Hey @codope After I apply the patch and rebuild the `trinobase` and `trinocoordinator`, the updated jar can successfully changed if I change the jar `docker/hoodie/hadoop/trinobase/target/hudi-trino-bundle.jar` and restart the image. I've commented how to test it in the pr https://github.com/apache/hudi/pull/9781.

codope commented:
@boneanxs Since #9781 landed, I have rebased this PR and removed the commented trino tests in `ITTestHoodieDemo`.

boneanxs commented:
> @boneanxs Since #9781 landed, I have rebased this PR and removed the commented trino tests in `ITTestHoodieDemo`.

Thanks @codope, the trino tests pass now.

boneanxs commented:
@hudi-bot run azure

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/rollback/BaseRollbackActionExecutor.java:
```diff
@@ -261,7 +261,10 @@ protected void finishRollback(HoodieInstant inflightInstant, HoodieRollbackMetad
-        table.getActiveTimeline().transitionRollbackInflightToComplete(inflightInstant,
+        // NOTE: no need to lock here, since !skipTimelinePublish is always true,
+        // when skipLocking is false, txnManager above-mentioned should lock it.
+        // when skipLocking is true, the caller should have already held the lock.
```
esp stuff like this, it's better to have re-entrancy here as opposed to avoiding locks, so we don't introduce code deps on locking, which is hard to debug.

danny0405 replied:
re-entrancy is good but you still need to keep track of the holdness of the lock explicitly, we have a `LockManager` in the `TxnManager`, but the `LockManager` is only available in the write client, and it is not designed as a singleton now.

vinothchandar commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/HoodieFlinkTableServiceClient.java:
```diff
@@ -137,6 +137,7 @@ protected void completeClustering(
 
       LOG.info("Committing Clustering {} finished with result {}.", clusteringCommitTime, metadata);
       table.getActiveTimeline().transitionReplaceInflightToComplete(
+          false,
```
why do we skip locking here completely.

danny0405 replied:
I guess it's because the file slicing is not affected by the replace operation, there is no need to keep atomicity.

vinothchandar commented on packaging/hudi-kafka-connect-bundle/pom.xml:
```diff
@@ -107,6 +107,7 @@
                                     <include>com.lmax:disruptor</include>
                                     <include>com.github.davidmoten:guava-mini</include>
                                     <include>com.github.davidmoten:hilbert-curve</include>
+                                    <include>com.github.ben-manes.caffeine:caffeine</include>
```
separate PR?

danny0405 replied:
+1, we can use a weak reference hash map now and we should add a close hook for the instances.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/config/HoodieTimeGeneratorConfig.java:
```diff
@@ -0,0 +1,128 @@
+ */
+public class HoodieTimeGeneratorConfig extends HoodieConfig {
+
+  public static final String BASE_PATH = "hoodie.base.path";
```
move this to `HoodieCommonConfig` and reference in `HoodieWriteConfig` and here?

danny0405 replied:
+1

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/config/HoodieTimeGeneratorConfig.java:
```diff
@@ -0,0 +1,128 @@
+      })
+      .sinceVersion("1.0.0")
+      .markAdvanced()
+      .withDocumentation("The max expected clock skew time for WaitBasedTimeGenerator");
```
The max expected clock skew time for WaitBasedTimeGenerator **in ms** ?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/config/HoodieTimeGeneratorConfig.java:
```diff
@@ -0,0 +1,128 @@
+
+  public static final ConfigProperty<Long> MAX_EXPECTED_CLOCK_SKEW_MS = ConfigProperty
+      .key("hoodie.time.generator.max_expected_clock_skew_ms")
+      .defaultValue(20L)
```
I'd say something like 200ms is a safer default.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimeGeneratorType.java:
```diff
@@ -0,0 +1,34 @@
+
+  @EnumFieldDescription("A wait based time generator, holding a mutex lock with skew times to "
+      + "produce globally monotonically increasing timestamp")
+  LOCK_PROVIDER
```
rename: WAIT_BASED ? Calling it `LOCK_PROVIDER` is a bit confusing IMO, since we have other meanings for this in the code/docs. 
In fact, call this sth like - `WAIT_TO_ADJUST_SKEW` to make it more explicit?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/config/LockConfiguration.java:
```diff
@@ -35,19 +35,23 @@ public class LockConfiguration implements Serializable {
   public static final String LOCK_ACQUIRE_RETRY_MAX_WAIT_TIME_IN_MILLIS_PROP_KEY = LOCK_PREFIX + "max_wait_time_ms_between_retry";
 
   public static final String LOCK_ACQUIRE_CLIENT_RETRY_WAIT_TIME_IN_MILLIS_PROP_KEY = LOCK_PREFIX + "client.wait_time_ms_between_retry";
+  public static final String DEFAULT_LOCK_ACQUIRE_CLIENT_RETRY_WAIT_TIME_IN_MILLIS = String.valueOf(5000L);
```
we probably need to turn these into ConfigProperty? (idk if there was a reason we could nt do this.) @yihua may know.

yihua replied:
Wrt this PR, is this refactoring existing code?  We have common lock configs in `HoodieLockConfig` class.

yihua replied:
@danny0405 @boneanxs ^

boneanxs replied:
`TimeGeneratorBase` need this default value, it cannot access `HoodieLockConfig` since it's not in `hudi-common`

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimeGeneratorBase.java:
```diff
@@ -0,0 +1,147 @@
+    return lockProvider;
+  }
+
+  public void lock() {
```
is there a common place to move this logic to? I guess this would be common to all locking code, right>?

danny0405 replied:
Yeah, I think we can have a refactoring to the failure retries code snippet, maybe we can abstrac it out into a utilitiy method.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimeGenerators.java:
```diff
@@ -0,0 +1,55 @@
+import org.apache.hudi.common.config.SerializableConfiguration;
+import org.apache.hudi.common.util.ValidationUtils;
+
+import com.github.benmanes.caffeine.cache.Cache;
```
is this a new dep?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -370,6 +377,28 @@ public synchronized HoodieActiveTimeline reloadActiveTimeline() {
+   *
+   * @param shouldLock whether the lock should be enabled to get the instant time.
+   */
+  public String createNewInstantTime(boolean shouldLock) {
```
I am thinking if we can simplify all the `shouldLock` passing somehow. Would it be possible to just have the lock configuration passed down and let the `TimeGenerator` decide? Default can be inprocess lock provider?

vinothchandar replied:
My worry is all the places we pass `true` `false` and how it may make it hard to maintain .

danny0405 replied:
We've been thinking through this for a while and we want to avoid the dead-lock, the lock providers other than in-process is kind of not reentrant now.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -198,8 +204,23 @@ private void readObject(java.io.ObjectInputStream in) throws IOException, ClassN
+   * @param shouldLock whether the lock should be enabled.
+   * @param instant the complete instant.
+   */
+  public void createCompleteInstant(boolean shouldLock, HoodieInstant instant) {
```
what are the cases where `shouldLock` is true vs false?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -447,37 +490,41 @@ public HoodieInstant transitionLogCompactionRequestedToInflight(HoodieInstant re
     ValidationUtils.checkArgument(inflightInstant.isInflight());
     HoodieInstant commitInstant = new HoodieInstant(State.COMPLETED, DELTA_COMMIT_ACTION, inflightInstant.getTimestamp());
-    transitionState(inflightInstant, commitInstant, data);
+    transitionStateToComplete(shouldLock, inflightInstant, commitInstant, data);
```
this special handling completed instant is because we only generate completion time for those?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -573,54 +624,85 @@ public HoodieInstant transitionReplaceRequestedToInflight(HoodieInstant requeste
+        HoodieInstant instantWithCompletionTime = new HoodieInstant(toInstant.getState(), toInstant.getAction(),
+            toInstant.getTimestamp(), metaClient.createNewInstantTime(false));
+        Path toInstantPath = getInstantFileNamePath(instantWithCompletionTime.getFileName());
+        boolean success = metaClient.getFs().rename(fromInstantPath, toInstantPath);
```
we should be avoiding renames, for cloud storage? I think we should make higher level APIs like 

HoodieStorage.writeFileAtomic(..)  and implement differently for hdfs and cloud storage say.

vinothchandar reviewed (changes requested):
This is great to see! Thanks for opening this. 

I focussed on `hudi-common` changes and have a bunch of questions around the locking configs. I am sure you have thought a bunch around this. So curious about what you think.

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -964,7 +976,7 @@ public void update(HoodieRestoreMetadata restoreMetadata, String instantTime) {
       // We need to choose a timestamp which would be a validInstantTime for MDT. This is either a commit timestamp completed on the dataset
       // or a timestamp with suffix which we use for MDT clean, compaction etc.
-      String syncCommitTime = HoodieTableMetadataUtil.createRestoreTimestamp(HoodieActiveTimeline.createNewInstantTime());
+      String syncCommitTime = HoodieTableMetadataUtil.createRestoreTimestamp(writeClient.createNewInstantTime(false));
```
Currently, only 2 use cases need a explicit lock:

1. create new instant time
2. transform pending metadata file to complete state

And if the caller already holds a lock, we must not use a lock to avoid dead-lock.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/fs/FSUtils.java:
```diff
@@ -187,7 +188,11 @@ public static String maskWithoutFileId(String instantTime, int taskPartitionId)
+    try {
+      return HoodieInstant.extractTimestamp(commitFileName);
+    } catch (IllegalArgumentException e) {
+      return "";
```
+1 to throw exception.

boneanxs commented on hudi-common/src/main/java/org/apache/hudi/client/transaction/lock/InProcessLockProvider.java:
```diff
@@ -19,16 +19,15 @@
 import org.apache.hudi.exception.HoodieLockException;
 
 import org.apache.hadoop.conf.Configuration;
-import org.jetbrains.annotations.NotNull;
```
Remove import `NotNull` here since hudi-common doesn't include dep `org.jetbrains.annotations`, which could meet ClassNotFound issue when compiling. Also, I see other lockProviders also don't use this annotation, like `ZookeeperBasedLockProvider` and `HiveMetastoreBasedLockProvider`, so simply remove it here.

danny0405 reviewed (approved):
Looks good overall, is there anyway we can abstract that failure retries as a common utility? Add add a UT for `TimeGenerator`.

boneanxs commented:
> Looks good overall, is there anyway we can abstract that failure retries as a common utility? Add add a UT for TimeGenerator.
done

waywtdcc commented:
![image](https://github.com/apache/hudi/assets/59957056/6cf60022-5d42-4e12-afd2-dac7a1f85e4b)
This is an error when upgrading from 0.12.1 to 1.0.0-beta1. Is this compatible with the file format of the old version?

danny0405 commented:
No, we would fix the ugrade for 1.0.0 GA release.

waywtdcc commented:
> No, we would fix the ugrade for 1.0.0 GA release.

Does this mean that the official version 1.0.0 will support automatic upgrade and fix this bug?

boneanxs commented:
> > No, we would fix the ugrade for 1.0.0 GA release.
> 
> Does this mean that the official version 1.0.0 will support automatic upgrade and fix this bug?

@waywtdcc Yeap, will fix it recently.

waywtdcc commented:
The meta file of version 1.0.0 is in avro format, and the old version is in json format. This should also be compatible.
